rust/kcl-lib/src/execution/types.rs

```diff
@@ -1178,26 +1189,55 @@ impl KclValue {
         exec_state: &mut ExecState,
         allow_shrink: bool,
     ) -> Result<KclValue, CoercionError> {
-        if len.satisfied(1, false).is_some() && self.has_type(ty) {
-            return Ok(KclValue::HomArray {
-                value: vec![self.clone()],
-                ty: ty.clone(),
-            });
-        }
         match self {
             KclValue::HomArray { value, ty: aty } => {
                 if aty.subtype(ty) {
-                    len.satisfied(value.len(), allow_shrink)
+                    // If the element type is a subtype of the target type and
+                    // the length constraint is satisfied, we can just return
+                    // the values unchanged, only adjusting the length. The new
+                    // array element type should preserve its type because the
+                    // target type oftentimes includes an unknown type as a way
+                    // to say that the caller doesn't care.
+                    return len
+                        .satisfied(value.len(), allow_shrink)
                         .map(|len| KclValue::HomArray {
                             value: value[..len].to_vec(),
                             ty: aty.clone(),
                         })
-                        .ok_or(self.into())
-                } else {
-                    Err(self.into())
+                        .ok_or(self.into());
+                }
+
+                // Ignore the array type, and coerce the elements of the array.
+                let mut values = Vec::new();
+                for item in value {
+                    if let KclValue::HomArray {
+                        ty: inner_ty,
+                        value: inner_value,
+                    } = item
+                    {
+                        // Flatten elements.
+                        if inner_ty.subtype(ty) {
```

Comment:
I think we should try to coerce here, not just subtype check since it's likely the array type is `any`.

Reply:
So I think the logic here is: if we have an array, check if the array can be coerced to the target type, if not, then flatten the array so we only flatten if we can't coerce without flattening, rather than try to flatten first. That means we can also try the coercions of non-array values in this loop rather than later and possibly fail quicker and avoid a clone

Reply:
Are you saying remove this entire block above so that we always coerce? `if aty.subtype(ty) { ... }` That's what I did.

Reply:
Nope. Omitting that broke `coerce_axes`.